Source/objects.cpp

```diff
@@ -2733,32 +2733,30 @@ void OperateShrineEnchanted(Player &player)
 	if (&player != MyPlayer)
 		return;
 
-	int cnt = 0;
-	uint64_t spell = 1;
-	int maxSpells = gbIsHellfire ? MAX_SPELLS : 37;
-	uint64_t spells = player._pMemSpells;
-	for (int j = 0; j < maxSpells; j++) {
-		if ((spell & spells) != 0)
-			cnt++;
-		spell *= 2;
-	}
-	if (cnt > 1) {
-		spell = 1;
+	const unsigned maxSpells = gbIsHellfire ? MAX_SPELLS : 37;
+
+	const uint64_t knownSpellMask = (0xFFFFFFFFULL >> (64 - maxSpells));
+	if ((player._pMemSpells & knownSpellMask) != 0) {
+		// If the player knows any spells
+		int r;
+		do {
+			// choose a spell at random to decrement
+			r = GenerateRnd(maxSpells) + 1;
+		} while ((player._pMemSpells & GetSpellBitmask(r)) == 0);
+
 		for (int j = SPL_FIREBOLT; j < maxSpells; j++) { // BUGFIX: < MAX_SPELLS, there is no spell with MAX_SPELLS index (fixed)

```

Comment:
The cnt loop and random index can both choose a spell in index maxSpells, this loop could result in adding a level to all known spells and not decrementing a level from any in that case.